File: exercism.io/ruby/difference-of-squares/difference_of_squares.rb
class Squares
  def initialize(n)
    @numbers = 1..n
  end

  def square_of_sum
    numbers.reduce(&:+)**2
  end

  def sum_of_squares
    numbers.map { |e| e**2 }.reduce(&:+)
  end

  def difference
    square_of_sum - sum_of_squares
  end

  private

  attr_reader :numbers
end

module BookKeeping
  VERSION = 4
end

File: exercism.io/erlang/difference-of-squares/src/difference_of_squares.erl
-module(difference_of_squares).

-export([difference_of_squares/1, square_of_sum/1, sum_of_squares/1]).


difference_of_squares(Number) ->
    square_of_sum(Number) - sum_of_squares(Number).

square_of_sum(Number) ->
    Numbers = lists:seq(1, Number),
    Sum = lists:foldl(fun(N, Sum) -> N + Sum end, 0, Numbers),
    trunc(math:pow(Sum, 2)).

sum_of_squares(Number) ->
    Numbers = lists:seq(1, Number),
    Squares = lists:map(fun(N) -> math:pow(N, 2) end, Numbers),
    Sum = lists:foldl(fun(N, Sum) -> N + Sum end, 0, Squares),
    trunc(Sum).
